edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/AuthAgentRequestHandler.cs

```diff
@@ -0,0 +1,377 @@
+
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Util;
+
+    using Microsoft.Extensions.DependencyInjection.Extensions;
+    using Microsoft.Extensions.Logging;
+    using Microsoft.Net.Http.Headers;
+
+    using Newtonsoft.Json;
+
+    public class AuthAgentRequestHandler
+    {
+        // The maximum allowed request size. The biggest item in a request is the certificate chain,
+        // 32k for the certificates and other data should be more than enough. Bigger requests are
+        // refused.
+        const int MaxInBufferSize = 32 * 1024;
+        const string Post = "POST";
+
+        static readonly TimeSpan requestTimeout = TimeSpan.FromSeconds(5);
+
+        readonly IAuthenticator authenticator;
+        readonly IUsernameParser usernameParser;
+        readonly IClientCredentialsFactory clientCredentialsFactory;
+        readonly AuthAgentProtocolHeadConfig config;
+
+        public AuthAgentRequestHandler(IAuthenticator authenticator, IUsernameParser usernameParser, IClientCredentialsFactory clientCredentialsFactory, AuthAgentProtocolHeadConfig config)
+        {
+            this.authenticator = Preconditions.CheckNotNull(authenticator, nameof(authenticator));
+            this.usernameParser = Preconditions.CheckNotNull(usernameParser, nameof(usernameParser));
+            this.clientCredentialsFactory = Preconditions.CheckNotNull(clientCredentialsFactory, nameof(clientCredentialsFactory));
+            this.config = Preconditions.CheckNotNull(config, nameof(config));
+        }
+
+        public static IWebHost CreateWebHostBuilder(
+                                    IAuthenticator authenticator,
+                                    IUsernameParser usernameParser,
+                                    IClientCredentialsFactory clientCredentialsFactory,
+                                    AuthAgentProtocolHeadConfig config)
+        {
+            return WebHost.CreateDefaultBuilder()
+                          .UseStartup<AuthAgentRequestHandler>()
+                          .UseUrls($"http://*:{config.Port}")
+                          .ConfigureServices(s => s.TryAddSingleton(authenticator))
+                          .ConfigureServices(s => s.TryAddSingleton(usernameParser))
+                          .ConfigureServices(s => s.TryAddSingleton(clientCredentialsFactory))
+                          .ConfigureServices(s => s.TryAddSingleton(config))
+                          .ConfigureLogging(c => c.ClearProviders())
+                          .Build();
+        }
+
+        public void Configure(IApplicationBuilder app) => app.Use(this.CallFilter).Run(this.Handler);
+
+        async Task Handler(HttpContext context)
+        {
+            try
+            {
+                Events.AuthRequestReceived();
+
+                var requestString = await ReadRequestBodyAsStringAsync(context.Request);
```

Comment:
Nit: please use actual types when the type is not clear..

Reply:
added type (and destroyed the harmony of identification 😒 )